includes/lib/glotpress-helper.php

```diff
@@ -0,0 +1,270 @@
+			if ( $type . 's' === $requested_type && $slug === $args['slug'] ) {
+				return get_translations( $type, $args['slug'], $api_url );
+			}
+
+			return $result;
+		},
+		10,
+		3
+	);
+
+	/**
+	 * Filters the translations transients to include the private plugin or theme.
+	 *
+	 * @see wp_get_translation_updates()
+	 */
+	add_filter(
+		'site_transient_update_' . $type . 's',
+		static function ( $value ) use ( $type, $slug, $api_url ) {
+			if ( ! $value ) {
+				$value = new \stdClass();
+			}
+
+			if ( ! isset( $value->translations ) ) {
+				$value->translations = [];
+			}
+
+			$translations = get_translations( $type, $slug, $api_url );
+
+			if ( ! isset( $translations['translations'] ) ) {
+				return $value;
+			}
+
+			$installed_translations = get_installed_translations( $type );
+			$locales                = get_available_locales();
+
+			/** This filter is documented in wp-includes/update.php */
+			$locales        = apply_filters( $type . 's_update_check_locales', $locales );
+			$active_locales = array_unique( $locales );
+
+			foreach ( (array) $translations['translations'] as $translation ) {
+				if ( ! \in_array( $translation['language'], $active_locales, true ) ) {
+					continue;
+				}
+
+				if ( $translation['updated'] && isset( $installed_translations[ $slug ][ $translation['language'] ] ) ) {
+					$local  = sanitize_date( $installed_translations[ $slug ][ $translation['language'] ]['PO-Revision-Date'] );
+					$remote = new DateTime( $translation['updated'] );
+
+					if ( $local >= $remote ) {
+						continue;
+					}
+				}
+
+				$translation['type'] = $type;
+				$translation['slug'] = $slug;
+
+				$value->translations[] = $translation;
+			}
+
+			return $value;
+		}
+	);
+}
+
+/**
+ * Registers actions for clearing translation caches.
+ *
+ * @since 1.1.0
+ */
+function register_clean_translations_cache() {
+	$clear_plugin_translations = static function() {
+		clean_translations_cache( 'plugin' );
+	};
+	$clear_theme_translations  = static function() {
+		clean_translations_cache( 'theme' );
+	};
+
+	add_action( 'set_site_transient_update_plugins', $clear_plugin_translations );
+	add_action( 'delete_site_transient_update_plugins', $clear_plugin_translations );
+
+	add_action( 'set_site_transient_update_themes', $clear_theme_translations );
+	add_action( 'delete_site_transient_update_themes', $clear_theme_translations );
+}
+
+/**
+ * Clears existing translation cache for a given type.
+ *
+ * @since 1.1.0
+ *
+ * @param string $type Project type. Either plugin or theme.
+ */
+function clean_translations_cache( $type ) {
+	$transient_key = constant( __NAMESPACE__ . '\TRANSIENT_KEY_' . strtoupper( $type ) );
+	$translations  = get_site_transient( $transient_key );
+
+	if ( ! \is_object( $translations ) ) {
+		return;
+	}
+
+	/*
+	 * Don't delete the cache if the transient gets changed multiple times
+	 * during a single request. Set cache lifetime to maximum 15 seconds.
+	 */
+	$cache_lifespan   = 15;
```

Comment:
Can we document how the caching works? Maybe WP core is doing it's own caching too.

How often will the site hit our servers? Ideally we don't want the admin pages to be hitting our servers every time they load a plugins page or our settings pages.

Reply:
We can get way more aggressive with the cache. I'd be okay to bump it up by a day or two (we have a way to manually download the translation as well or they could delete the transient to ping the server when needed).

Reply:
I spent some time looking into this. TLDR: our `clean_translations_cache()` here that clears caches older than 15 seconds should only be run twice daily (by default) on the core `set_site_transient_update_plugins` hook or whenever plugins are manually rechecked or a plugin is updated triggering the `delete_site_transient_update_plugins` hook.

So in a sense,  we do have some built-in caching from WP. It looks like the 15 second delay is really just there to prevent updating translations for a plugin multiple times on a single page load